presto-docs/src/main/sphinx/functions/geospatial.rst

```diff
@@ -144,7 +144,11 @@ Operations
 
     Returns a geometry that represents the point set union of the input geometries.
 
-    This function doesn't support geometry collections.
+.. function:: ST_Union(array<Geometry>) -> Geometry
```

Comment:
@tdcmeehan Tim,

SQL/MM spec defines `ST_Geometries(array<Geometry>)` function that constructs a geometry collection from an array of geometries. Would it make sense to add this function and use existing `ST_Union(Geometry)`? I think `ST_Geometries` is similar to PostGIS's `ST_Collect`.

<img width="843" alt="screen shot 2018-10-02 at 8 14 07 am" src="https://user-images.githubusercontent.com/27965151/46347909-29a69600-c61b-11e8-91c6-a6ed41f43f6e.png">

Reply:
@tdcmeehan Tim, today `ST_Union(ST_Geometries(ARRAY[...]))` would be less efficient than `ST_Union(ARRAY[...])` because of extra cost of serialization and deserialization of geometry returned by `ST_Geometries`. However, we discussed at some point to allow `Block` to store any object, not just `Slice`. Once we do that, `ST_Union(ST_Geometries(ARRAY[...]))` will be as efficient as `ST_Union(ARRAY[...])`. CC: @martint

Reply:
There are two components to this feedback, which have different impact on performance.

The first is to add `ST_Geometries` and have union operate on that.  I quickly wrote this up, for my JMH benchmark, I don't see any performance impact outside the margin of error:

```
Benchmark                                              Mode  Cnt   Score   Error  Units
BenchmarkGeometryAggregation.benchmarkGeometriesUnion  avgt   30  25.195 ± 0.584  ms/op
```

The second is to reuse the existing implementation of `ST_Union`.  What I presume you mean is overload `ST_Union` to accept just a single geometry, since the current implementation takes in two.  If I do that, and have it do something like take the first element of a geometry collection and union it with the entire geometry collection (reusing the existing `ST_Union` code), it's almost 50% slower.

```
Benchmark                                              Mode  Cnt   Score   Error  Units
BenchmarkGeometryAggregation.benchmarkGeometriesUnion  avgt   30  35.341 ± 1.170  ms/op
```

I suspect the reason for this is whenever the library detects a heterogenous union (either one or the other sides is a geometry collection, or the types of either side of the union don't match), it chooses a more robust algorithm to union, but which is slower.

Since there isn't much observed overhead by adding the `ST_Geometries`, what if I remove the public addition of the array form of `ST_Union`, but keep my private implementation in as shown in this PR, and add `ST_Geometries` so that users can still get this speedup, without us adding non-standard array form of `ST_Union`?

Reply:
@tdcmeehan Tim, sorry, I got confused. I was thinking of `ST_ConvexHull` which takes a single geometry and somehow forgetting that `ST_Union` takes two. I think `ST_Union` taking an array of geometries like you proposed initially makes sense. It is not part of the standard, but it is part of PostGIS and it makes sense generally. `ST_Union` taking a single geometry also not part of the standard, but is it not part of PostGIS either.

Let's go with `ST_Union(array<Geometry>)`. I'll review the changes.